### PR TITLE
Add new Indicator about buildings using ML

### LINKED ISSUE #243: Add new indicator for building completeness based on building area and ML
Waiting for:
- [x] #220 
- [ ] #229

### PR DESCRIPTION
**Pull request by matthiasschaub**
### Description

Predicts the building area of the AOI using a trained Random Forest Regressor.
The result is the ratio between the prediction of building area and the building
area mapped in OSM.

The input parameters (X or Covariates) to the models are population and population
density (GHSL GHS-POP), settlement typologies (GHSL SMOD), subnational Humand
Development Index (GDL SHDI) and nightlights (EGO VNL).

The spatial resolution of the model are hex-cells at zoom level 12. The input AOI is
split into hex-cells and the prediction is done for each of those hex-cells. The
model is trained on hex-cells in Africa. Therefor the Indicator is restricted to
input AOI within the bounding box of Africa.

### Corresponding issue
Closes #243 

### New or changed dependencies
- [building-completeness-model](https://gitlab.gistools.geog.uni-heidelberg.de/giscience/big-data/ohsome/ml-models/building-completeness-model/-/tree/main/)

### Checklist
- [x] I have updated my branch to `main` (e.g. through `git rebase main`)
- [x] My code follows the [style guide](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#style-guide) and was checked with [pre-commit](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CONTRIBUTING.md#pre-commit) before committing
- [x] I have commented my code
- [x] I have added sufficient unit and integration [tests](https://github.com/GIScience/ohsome-quality-analyst/blob/main/docs/development_setup.md#tests)
- [x] I have updated the [CHANGELOG.md](https://github.com/GIScience/ohsome-quality-analyst/blob/main/CHANGELOG.md)
- [x] Waiting for #227 
- [x] Waiting for #266 
- [x] Waiting for #272 
- [x] Waiting for #314 
- [x] Waiting for #333
- [x] Waiting for #334
- [x] Waiting for #335
- [x] Waiting for #353 
- [x] Check, whether raster containing no value and therefore indicator should be undefined
- [x] Create issue for: Possibilty for optimization if covariates (shdi, ghs-pop, etc.) are pre-computed for every hex-cell.
    - https://github.com/GIScience/ohsome-quality-analyst/issues/337

### CODE REVIEW

**Comment by matthiasschaub** [comment]
```
{
  "apiVersion": "0.9.0",
  "attribution": {
    "url": "https://github.com/GIScience/ohsome-quality-analyst/blob/main/data/COPYRIGHTS.md",
    "text": "© OpenStreetMap contributors"
  },
  "type": "Feature",
  "geometry": {
    "type": "MultiPolygon",
    "coordinates": [
      [
        [
          [
            5.779109,
            33.164272
          ],
          [
            5.779597,
            33.165833
          ],
          [
            5.873785,
            33.161541
          ],
          [
            5.953008,
            33.158386
          ],
          [
            5.994273,
            33.156437
          ],
          [
            6.00104,
            33.156494
          ],
          [
            6.003383,
            33.156322
          ],
          [
            6.010534,
            33.159302
          ],
          [
            6.032845,
            33.167492
          ],
          [
            6.050291,
            33.174145
          ],
          [
            6.064626,
            33.179127
          ],
          [
            6.065389,
            33.176319
          ],
          [
            6.065595,
            33.173397
          ],
          [
            6.06603,
            33.169556
          ],
          [
            6.06764,
            33.166409
          ],
          [
            6.069221,
            33.163601
          ],
          [
            6.070499,
            33.160736
          ],
          [
            6.071112,
            33.159073
          ],
          [
            6.071789,
            33.155117
          ],
          [
            6.07125,
            33.152485
          ],
          [
            6.07192,
            33.150482
          ],
          [
            6.075404,
            33.150593
          ],
          [
            6.076807,
            33.147846
          ],
          [
            6.076922,
            33.145035
          ],
          [
            6.07785,
            33.143661
          ],
          [
            6.078469,
            33.138214
          ],
          [
            6.079002,
            33.135067
          ],
          [
            6.079151,
            33.133175
          ],
          [
            6.079191,
            33.130653
          ],
          [
            6.077878,
            33.128132
          ],
          [
            6.077868,
            33.126587
          ],
          [
            6.077988,
            33.124237
          ],
          [
            6.077168,
            33.117935
          ],
          [
            6.080079,
            33.117073
          ],
          [
            6.075289,
            33.109226
          ],
          [
            6.072396,
            33.108307
          ],
          [
            6.063847,
            33.101608
          ],
          [
            6.062587,
            33.094959
          ],
          [
            6.057929,
            33.095989
          ],
          [
            6.049627,
            33.095417
          ],
          [
            6.046957,
            33.091751
          ],
          [
            6.045289,
            33.081322
          ],
          [
            6.044991,
            33.079891
          ],
          [
            6.019088,
            33.086136
          ],
          [
            6.001361,
            33.092896
          ],
          [
            5.900273,
            33.090488
          ],
          [
            5.778012,
            33.084068
          ],
          [
            5.758799,
            33.099312
          ],
          [
            5.779109,
            33.164272
          ]
        ]
      ]
    ]
  },
  "properties": {
    "metadata": {
      "name": "Building Area",
      "description": "Building Area"
    },
    "layer": {
      "name": "Building Area",
      "description": "All buildings as defined by all objects tagged with 'building=*'.\n"
    },
    "result": {
      "timestamp_oqt": "2022-05-24T08:25:01.333304+00:00",
      "timestamp_osm": "2022-05-15T20:00:00+00:00",
      "label": "red",
      "value": 0.1429655218503864,
      "description": "For the AOI the building area mapped in OSM is 957956.53 sqkm and\nthe predicted building area is 6700612.27 sqkm. The weighted\naverage of the ratio between the building area mapped in OSM and the\npredicted building area is 14.3 %. The weight is the\npredicted building area.\nThe building area mapped in OSM is significantly less than predicted.\nThis indicates that many buildings have not been mapped yet.\n"
    },
    "data": {
      "model_name": "Random Forest Regressor",
      "building_area_osm": [
        0,
        0,
        0,
        52486.14,
        841764.76,
        0,
        0,
        63705.63,
        0
      ],
      "building_area_prediction": [
        53320.37,
        5264.09625,
        57319.51,
        875144.24,
        3987139.72,
        840.03723294476,
        8152.7,
        1712591.56,
        840.03723294476
      ],
      "covariates": [
        {
          "ghs_pop": 0,
          "ghs_pop_density": 0,
          "water": 0,
          "very_low_density_rural": 1,
          "low_density_rural": 0,
          "rural_cluster": 0,
          "suburban_or_peri_urban": 0,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0,
          "urban_centre": 0,
          "shdi": 0.749352689479516,
          "vnl": 40.20217514038086
        },
        {
          "ghs_pop": 23.039753437042236,
          "ghs_pop_density": 2.4016750284244584e-7,
          "water": 0,
          "very_low_density_rural": 1,
          "low_density_rural": 0,
          "rural_cluster": 0,
          "suburban_or_peri_urban": 0,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0,
          "urban_centre": 0,
          "shdi": 0.749352689479516,
          "vnl": 3.270573616027832
        },
        {
          "ghs_pop": 0,
          "ghs_pop_density": 0,
          "water": 0,
          "very_low_density_rural": 1,
          "low_density_rural": 0,
          "rural_cluster": 0,
          "suburban_or_peri_urban": 0,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0,
          "urban_centre": 0,
          "shdi": 0.749352689479516,
          "vnl": 50.22721481323242
        },
        {
          "ghs_pop": 16318.097746707499,
          "ghs_pop_density": 0.000170098219228493,
          "water": 0,
          "very_low_density_rural": 0.83,
          "low_density_rural": 0.1,
          "rural_cluster": 0.02,
          "suburban_or_peri_urban": 0.01,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0.04,
          "urban_centre": 0,
          "shdi": 0.749352689479516,
          "vnl": 1710.898681640625
        },
        {
          "ghs_pop": 143944.96820783615,
          "ghs_pop_density": 0.0015004918277776384,
          "water": 0,
          "very_low_density_rural": 0.5591397849462365,
          "low_density_rural": 0.07526881720430108,
          "rural_cluster": 0,
          "suburban_or_peri_urban": 0.08602150537634409,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0,
          "urban_centre": 0.27956989247311825,
          "shdi": 0.749352689479516,
          "vnl": 9393.3876953125
        },
        {
          "ghs_pop": 0,
          "ghs_pop_density": 0,
          "water": 0,
          "very_low_density_rural": 1,
          "low_density_rural": 0,
          "rural_cluster": 0,
          "suburban_or_peri_urban": 0,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0,
          "urban_centre": 0,
          "shdi": 0.749352689479516,
          "vnl": 0
        },
        {
          "ghs_pop": 0,
          "ghs_pop_density": 0,
          "water": 0,
          "very_low_density_rural": 1,
          "low_density_rural": 0,
          "rural_cluster": 0,
          "suburban_or_peri_urban": 0,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0,
          "urban_centre": 0,
          "shdi": 0.749352689479516,
          "vnl": 4.482884883880615
        },
        {
          "ghs_pop": 62610.777252197266,
          "ghs_pop_density": 0.000652669167169395,
          "water": 0,
          "very_low_density_rural": 0.826530612244898,
          "low_density_rural": 0.061224489795918366,
          "rural_cluster": 0.02040816326530612,
          "suburban_or_peri_urban": 0,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0,
          "urban_centre": 0.09183673469387756,
          "shdi": 0.749352689479516,
          "vnl": 3988.3798828125
        },
        {
          "ghs_pop": 0,
          "ghs_pop_density": 0,
          "water": 0,
          "very_low_density_rural": 1,
          "low_density_rural": 0,
          "rural_cluster": 0,
          "suburban_or_peri_urban": 0,
          "semi_dense_urban_cluster": 0,
          "dense_urban_cluster": 0,
          "urban_centre": 0,
          "shdi": 0.749352689479516,
          "vnl": 0
        }
      ],
      "covariates_values": null,
      "hex_cell_geohash": [
        4171694,
        4172685,
        4171692,
        4170334,
        4170335,
        4172686,
        4171693,
        4170336,
        4172684
      ],
      "completeness_ratio": [
        0,
        0,
        0,
        0.059974273498046446,
        0.2111199554351208,
        0,
        0,
        0.03719837904608148,
        0
      ]
    }
  }
}
```

**Comment by Gigaszi** [comment]
> Does this resemble your implementation

Seems all right. The graph is less meaningful for small number of hexcells

File: workers/ohsome_quality_analyst/indicators/building_area/indicator.py [hunk 1]
```diff
@@ -0,0 +1,321 @@
...
+            get_raster_dataset("VNL"),
+            stats="sum",
+        )
+        smod_per_hex = get_smod_class_share(hex_cells)
+        shdi_per_hex = await db_client.get_shdi(hex_cells)
+        # Write covariates to a list of dataclasses (key-value pairs) for human readable
+        # output and to a nested list of the covariate values (scaled) for input to the
+        # model.
+        to_be_scaled = []
+        for hex_cell, ghs_pop, smod, shdi, vnl in zip(
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/indicators/building_area/indicator.py:138]: Changed to suggested syntax

File: workers/ohsome_quality_analyst/indicators/building_area/indicator.py [hunk 2]
```diff
@@ -0,0 +1,321 @@
...
+        scaled = min_max_scaler.transform(to_be_scaled)
+        for cov, s in zip(self.covariates_values, scaled):
+            cov[0] = s[0]  # ghs_pop
+            cov[1] = s[1]  # ghs_pop_density
+            cov[11] = s[2]  # vnl
+
+    def calculate(self) -> None:
+        # # Scale covariates
+        # Predict
+        random_forest_regressor = load_sklearn_model(
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/indicators/building_area/indicator.py:172]: Changed to `model`

File: workers/ohsome_quality_analyst/indicators/building_area/indicator.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        self.result.value = np.average(
+            self.completeness_ratio,
+            weights=self.building_area_prediction,
+        )
+        description = Template(self.metadata.result_description).substitute(
+            building_area_osm=round(sum(self.building_area_osm), 2),
+            building_area_prediction=round(sum(self.building_area_prediction), 2),
+            completeness_ratio=round(self.result.value * 100, 2),
+        )
+        if self.threshhold_green() <= self.result.value:
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/indicators/building_area/indicator.py:195]: Changed to suggested order

File: workers/ohsome_quality_analyst/indicators/building_area/metadata.yaml [hunk 4]
```diff
@@ -0,0 +1,26 @@
+---
+BuildingArea:
+  name: Building Area
+  description: |
+    Ratio of the predicted building area and the actual building area in OSM.
+    This can give an estimate if mapping has been completed.
+  label_description:
+    red: |
+      The building area mapped in OSM is significantly less than predicted.
+      This indicates that many buildings have not been mapped yet.
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/indicators/building_area/metadata.yaml:10]: changed

File: workers/ohsome_quality_analyst/indicators/building_area/metadata.yaml [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+    yellow: |
+      The building area mapped in OSM is less than predicted. This indicates
+      good amount of buildings but not all are already mapped.
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/indicators/building_area/metadata.yaml:13]: changed

File: workers/ohsome_quality_analyst/indicators/building_area/metadata.yaml [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    green: |
+      The building area mapped in OSM matches or exceeds the predicted building
+      area. This indicates good coverage of buildings are mapped in OSM.
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/ohsome_quality_analyst/indicators/building_area/metadata.yaml]: Grammar has been checked and is fixed

File: .pre-commit-config.yaml [hunk 7]
```diff
@@ -20,6 +20,7 @@ repos:
     hooks:
     -   id: check-added-large-files
         args: ["--maxkb=1024"]
+        exclude: "./workers/ohsome_quality_analyst/indicators/building_area/model.joblib"
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply .pre-commit-config.yaml:23]: Is removed

File: workers/tests/integrationtests/test_indicator_building_area.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        self.assertIsInstance(self.indicator.hex_cell_geohash, list)
+        self.assertGreater(len(self.indicator.hex_cell_geohash), 0)
+        # Covariates Values
+        self.assertIsNotNone(self.indicator.covariates_values)
+        self.assertIsInstance(self.indicator.covariates_values, list)
+        self.assertGreater(len(self.indicator.covariates_values), 0)
+        # Covariates
+        self.assertIsNotNone(self.indicator.covariates)
+        self.assertGreater(len(self.indicator.covariates), 0)
+        self.assertIsNotNone(self.indicator.covariates[0].ghs_pop)
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/tests/integrationtests/test_indicator_building_area.py]: Now a loop is used

File: workers/pyproject.toml [hunk 9]
```diff
@@ -30,7 +30,7 @@ python-dateutil = "^2.8.2"
 Jinja2 = "^3.0.3"
 scipy = "^1.7.3"  # `scipy` requires Python >=3.7,<3.11
 rpy2 = "^3.4.5"
-scikit-learn = "^1.0.2"  # Only works on Python below 3.11 (Depends on `scipy`)
+scikit-learn = "1.0.2"  # Only works on Python below 3.11 (Depends on `scipy`)
```
[the remark this answers is not in the document]
Comment (matthiasschaub) [reply workers/pyproject.toml]: Removed scikit-learn from dependencies

File: workers/pyproject.toml [hunk 10]
```diff
@@ -14,6 +14,11 @@ keywords = [
   "quality",
   ]
 
+[[tool.poetry.source]]
+name = "gistools"
```
[the remark this answers is not in the document]
Comment (Hagellach37) [reply workers/pyproject.toml:18]: shouldn't the name of the source be something like `building-completeness-model

**Review by Hagellach37 — APPROVED** [review]
looks good to me.

**Review by Hagellach37 — APPROVED** [review]

**Pull request merged**